pytables.ipynb

```diff
@@ -0,0 +1,287 @@
+      "  NaturalNameWarning)\n",
+      "/home/michael/miniconda3/envs/hetmech/lib/python3.6/site-packages/tables/path.py:112: NaturalNameWarning: object name is not a valid Python identifier: 'Gr>G_indices'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though\n",
+      "  NaturalNameWarning)\n",
+      "/home/michael/miniconda3/envs/hetmech/lib/python3.6/site-packages/tables/path.py:112: NaturalNameWarning: object name is not a valid Python identifier: 'Gr>G_indptr'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though\n",
+      "  NaturalNameWarning)\n",
+      "/home/michael/miniconda3/envs/hetmech/lib/python3.6/site-packages/tables/path.py:112: NaturalNameWarning: object name is not a valid Python identifier: 'Gr>G_shape'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though\n",
+      "  NaturalNameWarning)\n"
+     ]
+    }
+   ],
+   "source": [
+    "h5file = tables.open_file('data/sparse_matrices.h5', mode='w', title='Adjacency Matrices')\n",
+    "\n",
+    "for edge in abbr:\n",
+    "    group_path = f'{edge[0]}{edge[-1]}'\n",
+    "    try:\n",
+    "        h5file.create_group('/', group_path)\n",
+    "    except:\n",
+    "        pass\n",
+    "    r, c, sparse_matrix = metaedge_to_adjacency_matrix(graph, edge, sparse_threshold=1)\n",
+    "    group = h5file.create_group('/'+group_path+'/', edge)\n",
+    "    h5file.create_array(group, edge+'_data', sparse_matrix.data)\n",
+    "    h5file.create_array(group, edge+'_indices', sparse_matrix.indices)\n",
+    "    h5file.create_array(group, edge+'_indptr', sparse_matrix.indptr)\n",
+    "    h5file.create_array(group, edge+'_shape', sparse_matrix.shape)\n",
+    "\n",
+    "h5file.close()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 46,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "4.135847091674805 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "t1 = time.time()\n",
+    "h5file = tables.open_file('data/sparse_matrices.h5', 'a')\n",
+    "\n",
+    "node = 'CtD'\n",
+    "data = h5file.get_node(f'/{node[0]}{node[-1]}/{node}/{node}_data').read()\n",
+    "indices = h5file.get_node(f'/{node[0]}{node[-1]}/{node}/{node}_indices').read()\n",
+    "indptr = h5file.get_node(f'/{node[0]}{node[-1]}/{node}/{node}_indptr').read()\n",
+    "mat_shape = h5file.get_node(f'/{node[0]}{node[-1]}/{node}/{node}_shape').read()\n",
```

Comment:
This is a bit difficult to read. I think that one downside of the HDF5 approach is that it doesn't support sparse matrices as natively.

Reply:
Definitely a downside. The only reason I went about saving so many things was the hope that it could be significantly faster than other formats. If it were much faster, I'd probably just write `pt_read` and `pt_write` functions to make things more smooth with the several different arrays that need to be made for each sparse matrix.